### PR TITLE
Improve testing for temp directories

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Merge of PR #5483 to Jetty 10.0.x

### CODE REVIEW

**Comment by lachlan-roberts** [comment]
@janbartel this should be fixed now. The `WebAppProvider` works slightly differently in Jetty 10 as it uses a different thread to do the scan, and this broke the test.

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java [hunk 1]
```diff
@@ -165,8 +165,11 @@ public void resolveTempDirectory(WebAppContext context)
         //We need to make a temp dir. Check if the user has set a directory to use instead
         //of java.io.tmpdir as the parent of the dir
         File baseTemp = asFile(context.getAttribute(WebAppContext.BASETEMPDIR));
-        if (baseTemp != null && baseTemp.isDirectory() && baseTemp.canWrite())
+        if (baseTemp != null)
         {
+            if (!baseTemp.isDirectory() || !baseTemp.canWrite())
+                throw new IllegalStateException("BASETEMPDIR is not a writable directory");
```
Comment (janbartel) [inline jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java]:
```suggestion
                throw new IllegalStateException(WebAppContext.BASETEMPDIR + " is not a writable directory");
```

File: jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentTempDirTest.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    public void stop() throws Exception
+    {
+        server.stop();
+    }
+
+    @Test
+    public void testTmpDirectory() throws Exception
+    {
+        Path warPath = MavenTestingUtils.getTestResourcePath("webapps/foo-webapp-1.war");
+        String deploymentXml = "<!DOCTYPE Configure PUBLIC \"-//Jetty//Configure//EN\" \"https://www.eclipse.org/jetty/configure_9_3.dtd\">\n" +
```
Comment (janbartel) [inline jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentTempDirTest.java]: Should be configure_10_0.dtd?

File: jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentTempDirTest.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        // The temp directory has been cleared.
+        assertTrue(tmpDir2.exists());
+        assertThat(length(tmpDir2.listFiles()), is(0));
+    }
+
+    @Test
+    public void testPersistentTmpDirectory() throws Exception
+    {
+        Path warPath = MavenTestingUtils.getTestResourcePath("webapps/foo-webapp-1.war");
+        String deploymentXml = "<!DOCTYPE Configure PUBLIC \"-//Jetty//Configure//EN\" \"https://www.eclipse.org/jetty/configure_9_3.dtd\">\n" +
```
Comment (janbartel) [inline jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentTempDirTest.java]: Should be configure_10_0.dtd?

**Review by janbartel — CHANGES_REQUESTED** [review]

**Comment by lachlan-roberts** [comment]
@janbartel I have made the changes you requested.

**Comment by lachlan-roberts** [comment]
@janbartel nudge

**Review by janbartel — APPROVED** [review]
ok

**Pull request merged**